Allow minor individual regressions if overall test suite timing doesn't worsen

ZuleykhaPavlichenkova-TomTom opened the pull request:
Currently `regression/test_runner.py` sorts run results into two sections:
`REGRESSIONS DETECTED` and `OTHER TIMINGS`.

The overall test suite timing is also calculated, but it doesn't affect the decision whether there is a severe regression or not (the overall timings are printed together with the regression percentage).
If the geometric mean of new run is less and there are no individual regressions of +10%, then we could log these regressions as 'allowed', without reporting a regression, and don't fail the test run.

This PR updates `regression/test_runner.py` to allow minor individual regressions if overall test suite timing doesn't worsen by:
1. Adding a new category `ALLOWED REGRESSIONS` and print it in run results. 
2. Keeping the `REGRESSIONS DETECTED` section for the timings we'd like to report as an actual regressions.
3. Updating exit behaviour - `exit_code = 1` only if there are regressions to report. If only `allowed regressions` are present,  `exit_code = 0`.

This avoids noisy benchmark failures caused by insignificant regressions, while still reporting them clearly.
In the `FAILURES SUMMARY` section we will see both - `ALLOWED REGRESSIONS` and `REGRESSIONS DETECTED`

ZuleykhaPavlichenkova-TomTom committed:
If the geometric mean is less, then and there are no individual regressions of +10% I think we can safely report no regressions

ZuleykhaPavlichenkova-TomTom committed:
rearranged a report:
1. New category ALLOWED REGRESSIONS which we'd like to see in the test output as well
2. becnhmark goes to the REGRESSIONS DETECTED section only if it didn't go to the ALLOWED REGRESSIONS
3. error_code = 1 only when regressions to report list is not empty. When there are no regressions to report, but there are allowed regressions, then error_code = 0

ZuleykhaPavlichenkova-TomTom committed:
ff

Tmonster commented on scripts/regression/test_runner.py:
```diff
@@ -136,29 +136,86 @@ class BenchmarkResult:
-    exit_code = 1
-    print(
-        '''====================================================
+    print("HERE")
```
nit: looks like some debugging got left here?

hmeriann replied:
ah, yeah

Tmonster commented on scripts/regression/test_runner.py:
```diff
@@ -136,29 +136,86 @@ class BenchmarkResult:
+    print("HERE")
+    # filter regression_list to allowed and regressions to report
+    for regression in regression_list:
+        if isinstance(regression.old_result, (int, float)) or isinstance(regression.new_result, (int, float)):
```
Is this to catch regressions that fail? Is there an easier way to check this? Maybe `if regression.old_resuilt is not None`? 

And shouldn't we check both results? Make sure both are valid?

hmeriann replied:
> Is this to catch regressions that fail? Is there an easier way to check this? Maybe `if regression.old_resuilt is not None`?

Yes, and this part can be improved like this:
- We check if both results are numbers and we have max regression percentage defined, otherwise it goes to the report as it used to do

And I think I've found a better place to create this allowed regressions list

hmeriann replied:
Looking a bit more into the code, I see now that `regression-list` is already consists of the benchmarks regressed [to more than 10% ](https://github.com/duckdb/duckdb/blob/6e13b87b52176e2b598ea04d5b1044279e742c21/scripts/regression/test_runner.py#L132)(see [REGRESSION_THRESHOLD_PERCENTAGE](https://github.com/duckdb/duckdb/blob/6e13b87b52176e2b598ea04d5b1044279e742c21/scripts/regression/test_runner.py#L73C1-L73C32) value).
So we don't have to compare individual regressions to the `MAX_ALLOWED_REGRESS_PERCENTAGE` and, in case there are individual regressions AND overall regression exceeds max %, we can just replace `DETECTED REGRESSIONS` header with `ALLOWED REGRESSIONS` in that case

Tmonster commented on scripts/regression/test_runner.py:
```diff
@@ -136,29 +136,86 @@ class BenchmarkResult:
+            individual_regression_diff_perc = (
+                (regression.new_result - regression.old_result) / regression.old_result
+            ) * 100
+            if isinstance(time_old, float) and isinstance(time_new, float):
```
I think you can run this check once outside of the for loop

So outside the for loop something like this
```
if isinstance(time_old, float) and isinstance(time_new, float):
    gemean_time_slower = time_new <= time_old
```

Then inside

```
if gemean_time_slower and individual_regression_diff_perc <= 10.0:
```

And maybe the 10% can be a value we pass to the script with a default of 10. You can name it something like `MAX_ALLOWED_REGRESS_PERCENTAGE`

hmeriann replied:
oh yeah, this looks much cleaner and easier to read!
I'll add the MAX_ALLOWED_REGRESS_PERCENTAGE as an argument with no default value, because we don't set this boundary for the faster benchmarks to let them be considered as regression when they fail individually

Tmonster reviewed (commented):
Thanks!
Some questions.

ZuleykhaPavlichenkova-TomTom committed:
since the regression list already has all the benchmarks more than 10% regressed
we don't need to individually compare each of them to the max allowed percentage
and can just change the header based on the calculated regression percentage being
less than max percentage

regression percentage = int((time_new - time_old) * 100.0 / time_new)

default value for MAX_ALLOWED_REGRESS_PERCENTAGE=1

ZuleykhaPavlichenkova-TomTom committed:
ff

Tmonster commented on scripts/regression/test_runner.py:
```diff
@@ -65,14 +71,17 @@ def geomean(xs):
 # how many times we will run the experiment, to be sure of the regression
 NUMBER_REPETITIONS = 5
-# the threshold at which we consider something a regression (percentage)
+# the threshold at which we consider individual benchmark a regression (percentage)
```
you can also put this in the help

Tmonster commented on scripts/regression/test_runner.py:
```diff
@@ -141,59 +150,28 @@ class BenchmarkResult:
-    if len(allowed_regressions) > 0:
+    # regression_list already consists of the benchmarks regressed of more that 10%
+    regression_percentage = int((time_new - time_old) * 100.0 / time_new)
+    if isinstance(MAX_ALLOWED_REGRESS_PERCENTAGE, int) and regression_percentage < MAX_ALLOWED_REGRESS_PERCENTAGE:
```
why are we no longer looping through all regressions?

hmeriann replied:
To create a regression list we iterate over all the benchmarks and fall to three cases: 
- the benchmark run result is a string => it goes to the `error_list`
- the benchmark result on `new` build is slower => `regression_list`
- all other cases => `other_results`

To decide that `new` is closer, the runner does some [calculations](https://github.com/duckdb/duckdb/blob/6e13b87b52176e2b598ea04d5b1044279e742c21/scripts/regression/test_runner.py#L132), which does basically the the same as we did before and where, if we unfold all the values, we end up using `regression_threshold_percentage = 0.1`. So I thought that since we already check that the `old result + 0,05, 10% increased` is still less than a new result (which means a regression), and we added that benchmark to the `regression_list` we don't have to iterate over the `regression_list` comparing each value with the `MAX_ALLOWED_REGRESS_PERCENTAGE` and just make sure the overall geomean is bigger than `MAX_ALLOWED_REGRESS_PERCENTAGE`, to report all of them as regression.

Tmonster reviewed (commented):
Seems like we don't go through the regression list anymore? That doesn't make sense to me

ZuleykhaPavlichenkova-TomTom committed:
Update -help messages to become more informative

Tmonster commented on scripts/regression/test_runner.py:
```diff
@@ -47,7 +47,13 @@ def geomean(xs):
+    "--max-allowed-regression-percentage",
+    type=int,
+    default=1,
+    help="Max regression percentage for overal test suite to be a regression",
```
*overall

Tmonster replied:
I wonder if 1 is the correct value here. Can we see what the geomean time is from v1.1 to v1.2 and from v1.2 to v1.3? That can inform us a bit about what we are ok with/allow

hmeriann replied:
Will see - [here](https://github.com/hmeriann/nbc/actions/runs/15146070074/job/42581960862) I've triggered a run to check diffs with the TPCH large benchmarks
and [here](https://github.com/hmeriann/nbc/actions/runs/15146458242/job/42583211452) should also be a result for small benchmarks geomeans

Tmonster reviewed (commented):
Thanks!

I can understand why we just look at the geomean now. I'm worried now, however, that we run this more lenient check on every PR to duckdb/duckdb. I think we should stay a bit stricter there because the benchmarks we run there are smaller. 

Maybe we can add another flag so that this overall check is only performed when the WeeklyRegressionRunner checks?

hmeriann commented:
> Maybe we can add another flag so that this overall check is only performed when the WeeklyRegressionRunner checks?

If we pass the `MAX_ALLOWED_REGRESS_PERCENTAGE` to the `test_runner`, we can set a default value to 1, so we don't have to pass that value on regular tests, and we can use it instead of the flag (if `MAX_ALLOWED_REGRESS_PERCENTAGE + 1`).

And we can probably use it in calculations like [this](https://github.com/duckdb/duckdb/blob/ed6f6ea0780feee436cde01bc7e7a160b81c1206/scripts/regression/test_runner.py#L191)

ZuleykhaPavlichenkova-TomTom committed:
is_regression flag
the calsulation repeated in the end of run for readability

ZuleykhaPavlichenkova-TomTom committed:
ff

ZuleykhaPavlichenkova-TomTom committed:
add improved regressions list

ZuleykhaPavlichenkova-TomTom committed:
revert removed code

ZuleykhaPavlichenkova-TomTom committed:
ff

ZuleykhaPavlichenkova-TomTom committed:
add improvements list with improvement percentages

hmeriann commented:
Now it's also creates a list of improvements with individual improvements > 0%

ZuleykhaPavlichenkova-TomTom committed:
clean up benchmark data before rinnig CSV benchmarks

Tmonster commented on scripts/regression/test_runner.py:
```diff
@@ -133,32 +143,62 @@ class BenchmarkResult:
             regression_list.append(BenchmarkResult(benchmark, old_res, new_res))
         else:
-            other_results.append(BenchmarkResult(benchmark, old_res, new_res))
+            improved_by = int((new_res - old_res) * 100.0 / new_res)
```
If this is an improvement, wouldn't `new_res - old_res` always be negative? If a query goes from 10s to 8s, we would get 8 - 10 = -2.

Tmonster reviewed (commented):
Thanks! Looks like there is still merge conflict?

ZuleykhaPavlichenkova-TomTom committed:
Merge branch 'main' into allowed-regression

ZuleykhaPavlichenkova-TomTom committed:
some fixes

ZuleykhaPavlichenkova-TomTom committed:
Merge branch 'allowed-regression' of https://github.com/hmeriann/duckdb into allowed-regression

Tmonster commented on scripts/regression/test_runner.py:
```diff
@@ -151,32 +161,70 @@ class BenchmarkResult:
-    print(
-        '''====================================================
-==============  REGRESSIONS DETECTED   =============
+    if not (isinstance(time_old, str) or isinstance(time_new, str)):
```
can we also make sure `time_new > time_old`?
if `time_old > time_new` there are some regressions, but overall more improvements. Then you can just skip straight toward `regressions_header = 'ALLOWED_REGRESSIONS'`

Tmonster commented on scripts/regression/test_runner.py:
```diff
@@ -151,32 +161,70 @@ class BenchmarkResult:
+                summary.append(new_data)
+            print("")
+
+        if not (isinstance(time_old, str) or isinstance(time_new, str)):
```
will these values every be a string? If they are assigned on the lines
```
time_old = geomean(old_runner.complete_timings)
time_new = geomean(new_runner.complete_timings)
```
They should be doubles right? or ints?

Tmonster commented on scripts/regression/test_runner.py:
```diff
@@ -151,32 +161,70 @@ class BenchmarkResult:
+        and isinstance(regression_percentage, int)
+        and regression_percentage < MAX_ALLOWED_REGRESS_PERCENTAGE
+    ):
+        # allow individual regressions less than 10% when overall geomean had improved or hadn't change (on large benchmarks)
```
I'm not seeing the code where we go through the regressions_list to check each regression for less than 10%. Instead we are only checking the geomean. 

Am I missing the code for the individual checks somewhere?

Tmonster reviewed (commented):
Thanks!
Still a couple of questions
